Direct-XIP with BT OTA DFU docs (#12603)
nrfconnect/sdk-nrf: doc/nrf/device_guides/working_with_nrf/nrf52/developing.rst

FOTA over Bluetooth Low Energy
==============================

> Build configuration for MCUboot in the standard mode

Reviewer: Sorry, I changed my mind. There are lots of labels in this section that are potentially used in other Working with guides. Let's remove this subsection as it can destroy this composability (we introduce the notion of the MCUboot configuration that is not present in the  fota_upgrades_over_ble_intro)

Author: removed

Author: Discussed offline, will revert direct-xip mode and testing steps subsections and keep this one (standard mode) removed

.. fota_upgrades_over_ble_additional_information_end

.. fota_upgrades_over_ble_mcuboot_direct_xip_information_start

Build configuration additions for MCUboot in the direct-xip mode
----------------------------------------------------------------

FOTA updates are also supported when MCUboot is in the direct-xip mode.
In this mode, the MCUboot bootloader boots an image directly from a given slot, so the swap operation is not needed.
It can be used either with or without the revert mechanism support.
For more information about the direct-xip mode and the revert mechanism support, go to the Equal slots (direct-xip) section on the :doc:`mcuboot:design` page.

To use MCUboot in the direct-xip mode together with FOTA updates, do the following:

* Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP` Kconfig option in the MCUboot bootloader configuration.
* Enable the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP` Kconfig option to inform the application about the chosen MCUboot mode of operation.

>    * This Kconfig option automatically enables the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option to build application update images for both slots.

Reviewer: This part is the same in the case of the direct-xip mode with revert and without revert. Can we have a general comment after each mode description to avoid duplication? Any ideas @annwoj?

Author: Yeah, I wanted to include this note however i had no idea how to avoid writing it explicite.

Author: Replaced with one general note after everything, please take a look if it is acceptable for you.

     To see which files are built when the option is enabled, go to the :ref:`app_build_mcuboot_output` page.

See how to build the :ref:`peripheral_lbs` sample with MCUboot in the direct-xip mode when the revert mechanism support is disabled:

.. parsed-literal::
   :class: highlight

    west build -b *build_target* -- -Dmcuboot_CONFIG_BOOT_DIRECT_XIP=y -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_NCS_SAMPLE_MCUMGR_BT_OTA_DFU=y -DCONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP=y

Optionally, if you want to enable the revert mechanism support, complete the following:

* Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP_REVERT` Kconfig option in the MCUboot bootloader configuration.
* Enable the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP_WITH_REVERT` Kconfig option in place of :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP` to inform the application about the chosen MCUboot mode of operation.

>    * This Kconfig option automatically enables the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option to build application update images for both slots.

Reviewer: duplication from the previous comment

.. fota_upgrades_over_ble_additional_information_end

> FOTA updates are also supported when MCUboot is in Direct-XIP mode.

Reviewer: consider putting labels at the top and the bottom so that we can declare support for the direct-xip in case we add it for other platforms that are documented in their respective "Working with..." sections

In this mode, the MCUboot bootloader boots an image directly from a given slot, so the swap operation is not needed.
It can be used either with or without the revert mechanism support.

> For more information about Direct-XIP mode and the revert mechanism support, go to the :doc:`mcuboot:index-ncs` page.

Reviewer: Consider directly linking to the following section that describes the direct-xip mode:

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/mcuboot/design.html#equal-slots-direct-xip

Reviewer: @annwoj, is it possible to create such a reference?

Reviewer 2: We could at least try linking the bootloader page. There's a directive for that but it requires further "manual" referencing, because, as far as I know, there's no way to link the actual section through "ref". So we could go with something along the lines: (...) go to the Equal slots (direct-xip) section on the :doc:`Bootloader <mcuboot:design.md>` page.

To use MCUboot in Direct-XIP mode together with FOTA updates, do the following:

* Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP` Kconfig option in the MCUboot bootloader configuration.
* Enable the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option to build application update images for both slots.
  For more information, go to :ref:`app_build_mcuboot_output` page.

   * This Kconfig option automatically enables the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP` Kconfig option to inform the application about the chosen MCUboot mode of operation.
     Thus it is not required to set it manually if the Direct-XIP mode without the revert mechanism support is used.

* Optionally, if you want to enable the revert mechanism support, you also need to:

   * Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP_REVERT` Kconfig option in the MCUboot bootloader configuration.

>    * Enable the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP_WITH_REVERT` Kconfig option to inform the application about the chosen MCUboot mode of operation.

Reviewer: Do I understand correctly that the MCUmgr automatically confirms the image in the Direct-XIP mode like in the case of the standard (Swap) modes?

I just wanted to make sure that there is no need for the user to explicitly call the confirm API as it would require additional documentation.

.. parsed-literal::
   :class: highlight

    west build -b *build_target* -- -Dmcuboot_CONFIG_BOOT_DIRECT_XIP=y -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_NCS_SAMPLE_MCUMGR_BT_OTA_DFU=y -DCONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT=y

Here is an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode and the revert mechanism support:

.. parsed-literal::
   :class: highlight

    west build -b *build_target* -- -Dmcuboot_CONFIG_BOOT_DIRECT_XIP=y -Dmcuboot_CONFIG_BOOT_DIRECT_XIP_REVERT=y -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_NCS_SAMPLE_MCUMGR_BT_OTA_DFU=y -DCONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT=y -DCONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP_WITH_REVERT=y

.. note::
   When building the application for the first time with MCUboot in Direct-XIP mode and the revert mechanism support, you need to build the application with the additional option ``-DCONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS=\"--confirm\"``.
   This option will confirm the application during the image signing process.
   If the application is built without this option, it will fail to boot.

>    This option must be disabled when building update images.

Reviewer: It would be great to generate the flashed hex (`zephyr.hex`) with this option enabled and the `app_update.bin` with this option disabled. I am not sure if our build system supports it. Changing project configuration when you want to flash the DK or generate the DFU zip diminishes the UX. We can leave it for now, but it looks like a major thing for a future improvement

Reviewer: thanks

.. fota_upgrades_over_ble_additional_information_end

.. fota_upgrades_over_ble_mcuboot_direct_xip_information_start

> FOTA updates are also supported when MCUboot is in Direct-XIP mode.

Reviewer: I think it would be good to name this section. My suggestions:

"FOTA over Bluetooth Low Energy with MCUboot Direct-XIP mode"
or simply
"MCUboot Direct-XIP mode"

Reviewer: I think we can divide the FOTA over Bluetooth Low Energy section in the following subsections:

"Build configuration for MCUboot standard mode"

"Build configuration additions for MCUboot Direct-XIP mode"

"Testing steps"

You can create a separate commit for this

In this mode, the MCUboot bootloader boots an image directly from a given slot, so the swap operation is not needed.
It can be used either with or without the revert mechanism support.
For more information about Direct-XIP mode and the revert mechanism support, go to the :doc:`mcuboot:design` page.

To use MCUboot in Direct-XIP mode together with FOTA updates, do the following:

* Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP` Kconfig option in the MCUboot bootloader configuration.
* Enable the :kconfig:option:`CONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT` Kconfig option to build application update images for both slots.
  For more information, go to :ref:`app_build_mcuboot_output` page.

   * This Kconfig option automatically enables the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP` Kconfig option to inform the application about the chosen MCUboot mode of operation.

>      Thus it is not required to set it manually if the Direct-XIP mode without the revert mechanism support is used.

Reviewer 2: ```suggestion
     It is not required to set the option manually if you use the Direct-XIP mode without the revert mechanism support.
```

> * Optionally, if you want to enable the revert mechanism support, you also need to:

Reviewer 2: ```suggestion
* Optionally, if you want to enable the revert mechanism support, complete the following:
```

   * Enable the :kconfig:option:`CONFIG_BOOT_DIRECT_XIP_REVERT` Kconfig option in the MCUboot bootloader configuration.
   * Enable the :kconfig:option:`CONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP_WITH_REVERT` Kconfig option to inform the application about the chosen MCUboot mode of operation.

> Here is an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode without the revert mechanism support:

Reviewer 2: ```suggestion
See an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode that has disabled revert mechanism support:
```

Reviewer 2: Would it o be a good idea to move the example above the line 101? Since the option without revert mechanism support is discussed there

.. parsed-literal::
   :class: highlight

    west build -b *build_target* -- -Dmcuboot_CONFIG_BOOT_DIRECT_XIP=y -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_NCS_SAMPLE_MCUMGR_BT_OTA_DFU=y -DCONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT=y

> Here is an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode and the revert mechanism support:

Reviewer 2: ```suggestion
See an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode that has enabled revert mechanism support:
```

> Here is an example of how you can build for the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode and the revert mechanism support:

Reviewer 2: ```suggestion
See how to build the :ref:`peripheral_lbs` sample with MCUboot in Direct-XIP mode when the revert mechanism support is enabled:
```
as above - let me know if the meaning is affected

.. parsed-literal::
   :class: highlight

    west build -b *build_target* -- -Dmcuboot_CONFIG_BOOT_DIRECT_XIP=y -Dmcuboot_CONFIG_BOOT_DIRECT_XIP_REVERT=y -DCONFIG_BOOTLOADER_MCUBOOT=y -DCONFIG_NCS_SAMPLE_MCUMGR_BT_OTA_DFU=y -DCONFIG_BOOT_BUILD_DIRECT_XIP_VARIANT=y -DCONFIG_MCUBOOT_BOOTLOADER_MODE_DIRECT_XIP_WITH_REVERT=y

.. note::

>    When building the application for the first time with MCUboot in Direct-XIP mode and the revert mechanism support, you need to build the application with the additional option ``-DCONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS=\"--confirm\"``.

Reviewer 2: ```suggestion
   When building the application for the first time with MCUboot in Direct-XIP mode and the revert mechanism support, use an additional option ``-DCONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS=\"--confirm\"``.
```

>    This option will confirm the application during the image signing process.

Reviewer 2: ```suggestion
   This option will mark the application as ``confirmed`` during the image signing process.
```

   If the application is built without this option, it will fail to boot.

>    This option must be disabled when building update images.

Reviewer 2: ```suggestion
   It must, however, be disabled when building update images.
```

.. fota_upgrades_over_ble_additional_information_end

.. fota_upgrades_over_ble_mcuboot_direct_xip_information_start

Build configuration additions for MCUboot in the Direct-XIP mode
----------------------------------------------------------------

FOTA updates are also supported when MCUboot is in the Direct-XIP mode.
In this mode, the MCUboot bootloader boots an image directly from a given slot, so the swap operation is not needed.
It can be used either with or without the revert mechanism support.

> For more information about the Direct-XIP mode and the revert mechanism support, go to the :doc:`mcuboot:design` page.

Reviewer: ```suggestion
For more information about the Direct-XIP mode and the revert mechanism support, go to the Equal slots (direct-xip) section on the:doc:`mcuboot:design` page.
```

Reviewer 2: Yes, let's include the full reference just as per original suggestion

   |fota_upgrades_building|

#. Download the :file:`dfu_application.zip` zip archive or the :file:`app_update.bin` image file to your device.
   What file should be chosen depends on the application used to perform the firmware update.

>    See :ref:`app_build_fota` for more information about update files.

Reviewer: I think we should also modify the `app_build_fota` documentation section to cover the generated files in case of the Direct XIP mode (you can also add a note that the direct-xip is only supported on the single core devices - nRF52 series)

Author: I have noted in the build output files array that Direct-xip output is for single-core devices.

>    See :ref:`app_build_fota` for more information about update files.

Reviewer: I also think that we should update the table in the Multi-Image documentation:

https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/config_and_build/multi_image.html
